crates/net/nat/src/lib.rs

```diff
@@ -38,6 +39,72 @@ impl NatResolver {
     }
 }
 
+/// With this type you can resolve the external public IP address on an interval basis.
+#[must_use = "Does nothing unless polled"]
+pub struct ResolveNatInterval {
+    resolver: NatResolver,
+    future: Option<ResolveFut>,
+    interval: tokio::time::Interval,
+}
+
+// === impl ResolveNatInterval ===
+
+impl ResolveNatInterval {
+    fn with_interval(resolver: NatResolver, interval: tokio::time::Interval) -> Self {
+        Self { resolver, future: None, interval }
+    }
+
+    /// Creates a new [ResolveNatInterval] that attempts to resolve the public IP with interval of
+    /// period. See also [tokio::time::interval]
+    #[track_caller]
+    pub fn interval(resolver: NatResolver, period: Duration) -> Self {
+        let interval = tokio::time::interval(period);
+        Self::with_interval(resolver, interval)
+    }
+
+    /// Creates a new [ResolveNatInterval] that attempts to resolve the public IP with interval of
+    /// period with the first attempt starting at `sart`. See also [tokio::time::interval_at]
+    #[track_caller]
+    pub fn interval_at(
+        resolver: NatResolver,
+        start: tokio::time::Instant,
+        period: Duration,
+    ) -> Self {
+        let interval = tokio::time::interval_at(start, period);
+        Self::with_interval(resolver, interval)
+    }
+
+    /// Completes when the next [IpAddr] in the interval has been reached.
+    pub async fn tick(&mut self) -> Option<IpAddr> {
+        let ip = poll_fn(|cx| self.poll_tick(cx));
+        ip.await
+    }
+
+    /// Polls for the next resolved [IpAddr] in the interval to be reached.
+    ///
+    /// This method can return the following values:
+    ///
+    ///  * `Poll::Pending` if the next [IpAddr] has not yet been resolved.
+    ///  * `Poll::Ready(Option<IpAddr>)` if the next [IpAddr] has been resolved. This returns `None`
+    ///    if the attempt was unsuccessful.
+    pub fn poll_tick(&mut self, cx: &mut Context<'_>) -> Poll<Option<IpAddr>> {
```

Comment:
Why do we handroll this?

Reply:
are you referring to the function? this function mimics the interval API, but I'm not exactly sure what you mean by hand rolling here

Reply:
I meant why are we handrolling the future state mgmt instead of doing it as an async function, but maybe I'm misunderstanding

Reply:
this is actually pretty hard to do with async alone and would require some combination of stream! and select! macros, not pretty.